feature #4065 - gas and gas price can be changed for dapp transactions

goranjovic opened the pull request:
fixes #4065 

### Summary:

Adds the option to pass gas and gas limit for transactions created in dapps. The changes involve:

- `status-go` upgrade to `develop-g68966af0` for the dependency endpoints
- updated ios and android native modules to use the new endpoints
- updated UI to allow gas editing for dapp transactions

### Testing notes (optional):

Every send transaction option, not just from dapp (e.g. Wallet Send, chat command, both ether and tokens, etc.) - all where touched by the change.


### Steps to test:
- Send a transaction from dapp
- Tap Advanced, open gas settings and change them
- Send transaction and check the gas details in Transaction history and on Etherscan
- The same thing for other transaction flows

status: ready

goranjovic committed:
feature #4065 - gas and gas price can be changed for dapp transactions

Signed-off-by: Pedro Pombeiro <pombeirp@users.noreply.github.com>

flexsurfer commented on src/status_im/native_module/core.cljs:
```diff
@@ -40,8 +40,11 @@
+(defn approve-sign-request [id password callback]
+  (module-interface/-approve-sign-request rns-module id password callback))
+
+(defn approve-sign-request-with-args [id password gas gas-price callback]
```
why do we need two methods? maybe we could just set gas and gas-price `nil` and do not use `approve-sign-request` ?

goranjovic replied:
It's a constraint on go side and I prefer to keep 1:1 between go endpoints and native module functions.

flexsurfer commented on src/status_im/ui/screens/wallet/send/views.cljs:
```diff
@@ -106,6 +106,11 @@
 
+(defn return-to-transaction [modal?]
+  (if modal?
+    (re-frame/dispatch [:navigate-to-modal :wallet-send-transaction-modal])
```
custom navigation? something wrong, it shouldn't be possible to open modal from modal, we should discuss this with UX

goranjovic replied:
There's a good reason, let's discuss in chat.

mandrigin reviewed (approved):
LGTM

lukaszfryc commented:
For some reason which maybe unrelated to this PR, I can't send messages from iOS build:
<img src="https://user-images.githubusercontent.com/7532782/41166159-cf0cf4f4-6b3f-11e8-8a10-d8f22589d16b.png" width="200">

Everything works fine on Android (this PR build) or latest nightly (`03243e-n-fl`). @goranjovic is there any chance that this PR is related to this?

lukaszfryc commented:
Btw, I checked transactions in dapp (deploying a contract and getting STT in Sample DApp). That worked fine. I could my gas settings in transaction history and on etherscan. 

I still need to check wallet and chat transactions.

lukaszfryc commented:
@goranjovic one thing I've already noticed that we should fix is: Clicking on the "x" in transaction fee screen for a dapp closes the whole transaction instead of going back to the previous screen:

<img src="https://user-images.githubusercontent.com/7532782/41166924-c4e1255c-6b41-11e8-8b2c-11903662e38c.png" width="200">

Steps:
1. Open Sample Dapp
2. Click deploy contract to sign the transaction
3. Open transaction fee screen
4. Click "x" to go back to the previous screen
-> You end up in the dapp instead of sign transaction screen

lukaszfryc commented:
@goranjovic I've tested the latest build, here is what I found:

1) I still have the "x" issue described in the previous comment.

2) Also, I see that default gas price and limit are not shown in transaction fee screen. 
<img src="https://user-images.githubusercontent.com/7532782/41221826-99e22c9a-6d65-11e8-8b0b-0a9333eec449.png" width="200">

What's more, when I set these values manually and click done button, the sign transaction screen still shows the old values
<img src="https://user-images.githubusercontent.com/7532782/41221940-f1bbbc10-6d65-11e8-8166-a40892618c7e.png" width="200">

3) If I set too low gas in a dapp (for instance Airswap) I get the "intrinsic gas to low" error which is OK. But, when I go to the transaction history, I'm getting a BigNumber error:
<img src="https://user-images.githubusercontent.com/7532782/41222403-0903288a-6d67-11e8-8807-4600a4b2962d.png" width="200">

goranjovic commented:
@lukaszfryc Ok, but is chat still broken?

I was unsuccessfully trying to reproduce the transaction history error (3) over the weekend for another bugfix. This is mainnet right?

lukaszfryc commented:
> Ok, but is chat still broken?

No, all transactions should work fine now.

Yes, it's mainnet. I'm able to reproduce it 2/2 for Airswap

goranjovic commented:
@lukaszfryc Still can't reproduce the transaction history, but see a different problem. Are you able to Sign messages? I'm getting a number format exception (both Simple dapp on Ropsten and Airswap on Mainnet)

lukaszfryc commented:
@goranjovic I was able to sign them when I tested it couple of hours ago.

lukaszfryc commented:
@goranjovic re transaction history, let me try it again on a new build.

goranjovic commented:
@lukaszfryc re re transaction history, what keeps happening to me is that transaction with low gas never even appears in history, but no errors. Just to clarify, are you setting a low gas limit, gas price or both?

goranjovic commented:
@lukaszfryc Found the cause for the crash I was seeing on Sign message. Really weird it doesn't happen for you. Fixed it and will push along with close with X as soon as I address the empty gas values issue.

goranjovic commented:
@lukaszfryc Ready to be re-tested. I addressed all points except (3) `If I set too low gas in a dapp...`  which I still can't reproduce. 
That issue I'm addressing in #4712  so if you notice anything around that please comment there.

lukaszfryc commented:
@goranjovic 1, 2 work fine but there are 2 different issues now that occur only on Android:

1) Android status bar turns white when Transaction Fee for signing a dapp transaction is shown: 
<img src="https://user-images.githubusercontent.com/7532782/41280275-d6612e40-6e2e-11e8-8e8a-d92efb26bbc2.png" width="200">
It looks fine in chat transactions:
<img src="https://user-images.githubusercontent.com/7532782/41280527-6e0b965e-6e2f-11e8-9b6f-1f8ac3c096db.png" width="200">

2) I now can reproduce the format exception when trying to request token or sign a contract in Sample Dapp. The error shows after providing a password and clicking "Sign transaction". Everything works fine for other dapps like Airswap or for chat transactions.   
```
{"timestamp":1528793123,"message":"java.lang.NumberFormatException: null","stackTrace":"java.lang.NumberFormatException: null\n\tat java.lang.Long.parseLong(Long.java:406)\n\tat java.lang.Long.parseLong(Long.java:485)\n\tat im.status.ethereum.module.StatusModule$9.run(StatusModule.java:535)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)\n\tat java.lang.Thread.run(Thread.java:762)\n"}
```

goranjovic commented:
@lukaszfryc Fixed and pushed. Good catch for status bar, it was actually broken in Chat Send as well, just wasn't visible because the screen behind it was also blue.

lukaszfryc commented:
@goranjovic works great 👍But, can you fix "the [BigNumber error](https://github.com/status-im/status-react/pull/4682#issuecomment-396173830) in transaction history" in scope of this PR? Otherwise we're going to introduce this issue in `develop`.

goranjovic commented:
@lukaszfryc That issue wasn't introduced in this PR and I'd rather not expand the scope too much, this is a really big PR as it is. And besides, that bug has its own PR already.

lukaszfryc commented:
@goranjovic I'm only worried that by allowing users to set custom gas in dapps, they may encounter this issue more often. I could not reproduce it with wallet or chat transactions in recent nightly.

goranjovic commented:
@lukaszfryc And I could not reproduce it at all. When I try it the transaction just stays at 0 confirmations, which is expected - if the gas price is too low, nobody will want to mine it, fair enough. But I never see the error in transaction history.
Seeing the same behavior for chat/wallet transactions in the other PR and here.
What are the exact steps you do when you get this? Maybe I'm trying something slightly different.

lukaszfryc commented:
@goranjovic I can't reproduce it now either. Previously when it happened I set low gas and got the intrinsict gas too low. Then when I went to the transaction history I got the error popup. Anyway, I think we can create a separate issue for it.

But please give me ~20 more min, I still want to check few things.
